### ISSUE TITLE
coherence-operator deletion process hangs if PODs are in ImagePullBackOff status

### ISSUE THREAD

**Issue by kviktorman** [oracle/coherence-operator#474]
### Behavior

As I deployed the custom images by accident I forgot to define the same namespace where the coherence-operator and the image pull secret were installed. As a result the custom images were deployed to the default namespace and ended up in `ImagePullBackOff` status. When I tried to remove them the deletion process got stuck. 

We managed to overcome the stuck by executing the following command
```
# The "/cache-storage" is coming from the metadata name
$ kubectl patch coherence.coherence.oracle.com/cache-storage -p '{"metadata":{"finalizers":[]}}' --type=merge
```

Can you please have a look to the deletion process?

Thanks, 
Viktor

### Environment details

| Component | Version
| --- | --- |
| Coherence operator | 3.1.3 |
| Kubernetes | v1.15.3 |

### Used settings

```
# content of the cache-storage.yaml file

# below $IMAGE_REGISTRY is the custom URL of the registry
# below $SECRET_NAME is the custom secret used for image pulling

apiVersion: coherence.oracle.com/v1
kind: Coherence
metadata:
  name: cache-storage
spec:
  cluster: cache-cluster
  image: $IMAGE_REGISTRY
  imagePullSecrets:
    - name: $SECRET_NAME
  coherence:
    cacheConfig: /opt/coherence-cluster-server/coherence-configuration/coherence-server-cache-config.xml
    storageEnabled: true
    logLevel: 9
  jvm:
    useJibClasspath: false
    classpath:
      - "/opt/coherence-cluster-server/jars/*"       
```

### Operator logs

```
2021-02-03T15:12:52.129Z	INFO	controllers.Coherence	Finalizing Coherence resource	{"Namespace": "default", "Name": "cache-storage"}
2021-02-03T15:12:52.129Z	INFO	controllers.StatefulSet	Suspending Coherence services in StatefulSet cache-storage	{"Namespace": "default", "Name": "cache-storage"}
2021-02-03T15:12:52.129Z	INFO	controllers.StatefulSet	Cannot execute probe, one or more Pods is not in a ready state	{"Namespace": "default", "Name": "cache-storage"}
2021-02-03T15:12:52.129Z	ERROR	controller	Reconciler error	{"reconcilerGroup": "coherence.oracle.com", "reconcilerKind": "Coherence", "controller": "coherence", "name": "cache-storage", "namespace": "default", "error": "failed to suspend services"}
github.com/go-logr/zapr.(*zapLogger).Error
	/home/runner/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
```

### Steps to Reproduce

```
#1. add the coherence custom images 
$ kubectl apply -f cache-storage.yaml

#2. wait until the pods status gets to ImagePullBackOff
$ kubectl get pods
NAME              READY   STATUS             RESTARTS   AGE
cache-storage-0   0/1     ImagePullBackOff   0          6m19s
cache-storage-1   0/1     ImagePullBackOff   0          6m19s
cache-storage-2   0/1     ImagePullBackOff   0          6m19s

#3. try to remove the coherence custom images
# the deletion process will stuck
$ kubectl delete -f cache-storage.yaml
coherence.coherence.oracle.com "cache-storage" deleted  
_ 
```

**Comment by aseovic** [comment]
@kviktorman You can usually get "unstuck" by scaling the misconfigured cluster down to zero, after which the delete should succeed:
```
$ kubectl scale coherence cache-storage --replicas 0
```

**Comment by kviktorman** [comment]
@aseovic it seems the scale down will not solve this issue. It will give back the prompt after execution, but the PODs will stay.

In the operator log I can see this:
```
2021-02-04T08:41:57.625Z	ERROR	controller	Reconciler error	{"reconcilerGroup": "coherence.oracle.com", "reconcilerKind": "Coherence", "controller": "coherence", "name": "cache-storage", "namespace": "default", "error": "failed to suspend services prior to scaling down to zero"}
github.com/go-logr/zapr.(*zapLogger).Error
	/home/runner/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
```

**Comment by thegridman** [comment]
This is a known issue  and removing the finaliser as you did is currently the only way to fix it.

**Comment by kviktorman** [comment]
I'm wondering if it would be worth to have something about this in your [troubleshooting guide](https://oracle.github.io/coherence-operator/docs/latest/#/troubleshooting/01_trouble-shooting). What do you think?

**Comment by thegridman** [comment]
It was actually added in there a couple of days ago https://github.com/oracle/coherence-operator/blob/master/docs/troubleshooting/01_trouble-shooting.adoc but was too late to get into the 3.1.3 released docs. I could possibly manually update the released docs.

This scenario is probably one we could do something about as we can tell that none of the Pods are ready so it is safe to delete the deployment. The reason we have the finalizer is to do a clean safe shutdown without data loss, which in this scenario does not apply. I'll add it to our list of things for the next release.

**Comment by thegridman** [comment]
That's JIRA COH-22927 for anyone internal to Oracle who want's to know it's progress.

**Comment by thegridman** [comment]
Fixed in 3.1.4
